lib/varnishcfg/backends.go

```diff
@@ -0,0 +1,172 @@
+		conditions = append(conditions, fmt.Sprintf(`req.http.host == "%s"`, fqdn))
+	}
+
+	lines = append(lines, fmt.Sprintf("if (%s) {", strings.Join(conditions, " || ")))
+	lines = append(lines, fmt.Sprintf("\tset req.backend_hint = %s.backend();", svc.Name))
+
+	// only change request host from edge servers which typically has multiple request FQDNs or
+	// one request FQDN that is not the origin.
+	if len(requestFQDNs) > 1 || (len(requestFQDNs) == 1 && requestFQDNs[0] != svc.DestDomain) {
+		lines = append(lines, fmt.Sprintf("\tset req.http.host = \"%s\";", svc.DestDomain))
+	}
+
+	lines = append(lines, "}")
+
+	if _, ok := subroutines["vcl_recv"]; !ok {
+		subroutines["vcl_recv"] = make([]string, 0)
+	}
+	subroutines["vcl_recv"] = append(subroutines["vcl_recv"], lines...)
+}
+
+func addBackends(backends map[string]backend, parents []*atscfg.ParentAbstractionServiceParent, originDomain string, originPort int) {
+	for _, parent := range parents {
+		backendName := fmt.Sprintf("%s", getBackendName(parent.FQDN, parent.Port))
+		if _, ok := backends[backendName]; ok {
+			continue
+		}
+		backends[backendName] = backend{
+			host: parent.FQDN,
+			port: parent.Port,
+		}
+	}
+	backendName := fmt.Sprintf("%s", getBackendName(originDomain, originPort))
+	if _, ok := backends[backendName]; ok {
+		return
+	}
+	backends[backendName] = backend{
+		host: originDomain,
+		port: originPort,
+	}
+}
+
+func addDirectors(subroutines map[string][]string, svc *atscfg.ParentAbstractionService) {
+	lines := make([]string, 0)
+	fallbackDirectorLines := make([]string, 0)
+	fallbackDirectorLines = append(fallbackDirectorLines, fmt.Sprintf("new %s = directors.fallback();", svc.Name))
+
+	if len(svc.Parents) != 0 {
+		lines = append(lines, addBackendsToDirector(svc.Name+"_primary", svc.RetryPolicy, svc.Parents)...)
+		fallbackDirectorLines = append(fallbackDirectorLines, fmt.Sprintf("%s.add_backend(%s_primary.backend());", svc.Name, svc.Name))
+	}
+	if len(svc.SecondaryParents) != 0 {
+		lines = append(lines, addBackendsToDirector(svc.Name+"_secondary", svc.RetryPolicy, svc.SecondaryParents)...)
+		fallbackDirectorLines = append(fallbackDirectorLines, fmt.Sprintf("%s.add_backend(%s_secondary.backend());", svc.Name, svc.Name))
+	}
+	fallbackDirectorLines = append(fallbackDirectorLines, fmt.Sprintf("%s.add_backend(%s);", svc.Name, getBackendName(svc.DestDomain, svc.Port)))
+
+	lines = append(lines, fallbackDirectorLines...)
+
+	if _, ok := subroutines["vcl_init"]; !ok {
+		subroutines["vcl_init"] = make([]string, 0)
+	}
+	subroutines["vcl_init"] = append(subroutines["vcl_init"], lines...)
+}
+
+func addBackendsToDirector(name string, retryPolicy atscfg.ParentAbstractionServiceRetryPolicy, parents []*atscfg.ParentAbstractionServiceParent) []string {
+	lines := make([]string, 0)
+	directorType := getDirectorType(retryPolicy)
+	sticky := ""
+	if directorType == "fallback" && retryPolicy == atscfg.ParentAbstractionServiceRetryPolicyLatched {
```

Comment:
Consider adding a separate `sticky` return from getDirectorType so all that evaluation happens in the same place